### ISSUE TITLE
No tests from repo root or modules are run

### ISSUE THREAD

**Issue by greenatatlassian** [cekit/cekit#780]
**Describe the bug**
Test files from `tests` directory next to the module descriptor are ignored.

The problem is that the code in `collector.py` copies from the target directory, `target/images/modules/<module_name>/tests/*` as confirmed by `test_collect_test_from_module`. However, this directory has already been cleaned up by `cli.cleanup` which is called when the application starts up.

**Expected behavior**
I would have expected the tests to be copied into the target directory and run along side tests from other locations as part of `cekit test behave`.

**Logs**
```
> cat image.yaml 
───────┬────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
       │ File: image.yaml
───────┼────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
   1   │ schema_version: 1
   2   │ 
   3   │ name: "dummy"
   4   │ version: "1"
   5   │ description: "Golden Image with OS base layer for development based on Debian."
   6   │ 
   7   │ from: "debian:bullseye-slim"
   8 + │ 
   9 + │ modules:
  10 + │   repositories:
  11 + │     - path: "./modules"
  12 + │ 
  13 + │   install:
  14 + │     - name: "mymodule"
───────┴────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────

 > tree .
.
├── image.yaml
└── modules
    └── mymodule
        ├── module.yaml
        └── tests
            ├── features
            │   └── mymodule.feature
            └── steps
                └── steps.py

5 directories, 4 files

 > cat modules/mymodule/tests/features/mymodule.feature 
───────┬────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
       │ File: modules/mymodule/tests/features/mymodule.feature
───────┼────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
   1   │ Feature: my module
   2   │ 
   3   │   @dummy
   4   │   Scenario: Test features in module
   5   │     Given behave test framework with custom test
   6   │     Then test passes
───────┴────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────

 > cekit build docker
2022-05-23 09:07:09,718 cli.py:418        WARNING You are running unreleased development version of CEKit, use it only at your own risk!
2022-05-23 09:07:09,719 tools.py:330        WARNING You are running CEKit on an unknown platform. External dependencies suggestions may not work!
2022-05-23 09:07:09,719 tools.py:500        WARNING The certifi library (https://certifi.io/) was found, depending on the operating system configuration this may result in certificate validation issues
2022-05-23 09:07:09,719 tools.py:506        WARNING Certificate Authority (CA) bundle in use: '/Users/agreen/code/cekit-golden-images/cekit/lib/python3.9/site-packages/certifi/cacert.pem'
2022-05-23 09:07:09,884 builder.py:83         INFO  Generating files for docker engine
2022-05-23 09:07:09,884 base.py:84         INFO  Initializing image descriptor...
2022-05-23 09:07:09,887 resource.py:160        WARNING No value found for 'name' in '{"dest": "/tmp/artifacts/", "path": "./modules"}' artifact; using auto-generated value of 'modules'
2022-05-23 09:07:09,890 resource.py:208        INFO  Copying resource 'modules'...
2022-05-23 09:07:09,904 docker.py:25         INFO  Handling artifacts for docker...
2022-05-23 09:07:09,907 base.py:350        INFO  Rendering Dockerfile...
2022-05-23 09:07:09,959 docker_builder.py:251        INFO  Building container image...
2022-05-23 09:07:10,188 docker_builder.py:103        INFO  Docker: Step 1/7 : FROM debian:bullseye-slim
2022-05-23 09:07:10,189 docker_builder.py:103        INFO  Docker: 
2022-05-23 09:07:10,189 docker_builder.py:103        INFO  Docker: ---> ed595c52174a
2022-05-23 09:07:10,189 docker_builder.py:103        INFO  Docker: Step 2/7 : USER root
2022-05-23 09:07:10,189 docker_builder.py:103        INFO  Docker: 
2022-05-23 09:07:10,220 docker_builder.py:103        INFO  Docker: ---> Running in a663908811f0
2022-05-23 09:07:10,329 docker_builder.py:103        INFO  Docker: Removing intermediate container a663908811f0
2022-05-23 09:07:10,330 docker_builder.py:103        INFO  Docker: ---> 0b9a64a05ae7
2022-05-23 09:07:10,330 docker_builder.py:103        INFO  Docker: Step 3/7 : LABEL             description="Golden Image with OS base layer for development based on Debian."             io.cekit.version="4.2.0.dev0"             summary="Golden Image with OS base layer for development based on Debian."
2022-05-23 09:07:10,330 docker_builder.py:103        INFO  Docker: 
2022-05-23 09:07:10,361 docker_builder.py:103        INFO  Docker: ---> Running in de3d122bb1fd
2022-05-23 09:07:10,464 docker_builder.py:103        INFO  Docker: Removing intermediate container de3d122bb1fd
2022-05-23 09:07:10,464 docker_builder.py:103        INFO  Docker: ---> c608cec0da54
2022-05-23 09:07:10,464 docker_builder.py:103        INFO  Docker: Step 4/7 : USER root
2022-05-23 09:07:10,464 docker_builder.py:103        INFO  Docker: 
2022-05-23 09:07:10,490 docker_builder.py:103        INFO  Docker: ---> Running in 203554cfbbea
2022-05-23 09:07:10,598 docker_builder.py:103        INFO  Docker: Removing intermediate container 203554cfbbea
2022-05-23 09:07:10,598 docker_builder.py:103        INFO  Docker: ---> 5564f661444e
2022-05-23 09:07:10,598 docker_builder.py:103        INFO  Docker: Step 5/7 : RUN [ ! -d /tmp/scripts ] || rm -rf /tmp/scripts
2022-05-23 09:07:10,598 docker_builder.py:103        INFO  Docker: 
2022-05-23 09:07:10,622 docker_builder.py:103        INFO  Docker: ---> Running in ef3d1263c0a7
2022-05-23 09:07:11,178 docker_builder.py:103        INFO  Docker: Removing intermediate container ef3d1263c0a7
2022-05-23 09:07:11,178 docker_builder.py:103        INFO  Docker: ---> c4f716886f2c
2022-05-23 09:07:11,178 docker_builder.py:103        INFO  Docker: Step 6/7 : RUN [ ! -d /tmp/artifacts ] || rm -rf /tmp/artifacts
2022-05-23 09:07:11,179 docker_builder.py:103        INFO  Docker: 
2022-05-23 09:07:11,206 docker_builder.py:103        INFO  Docker: ---> Running in 518ac775a533
2022-05-23 09:07:11,753 docker_builder.py:103        INFO  Docker: Removing intermediate container 518ac775a533
2022-05-23 09:07:11,753 docker_builder.py:103        INFO  Docker: ---> 9ebd8d48c523
2022-05-23 09:07:11,753 docker_builder.py:103        INFO  Docker: Step 7/7 : USER root
2022-05-23 09:07:11,753 docker_builder.py:103        INFO  Docker: 
2022-05-23 09:07:11,785 docker_builder.py:103        INFO  Docker: ---> Running in cb0b19147ff0
2022-05-23 09:07:11,894 docker_builder.py:103        INFO  Docker: Removing intermediate container cb0b19147ff0
2022-05-23 09:07:11,894 docker_builder.py:103        INFO  Docker: ---> f1db17598144
2022-05-23 09:07:11,895 docker_builder.py:103        INFO  Docker: Successfully built f1db17598144
2022-05-23 09:07:11,895 docker_builder.py:157        INFO  Squashing image f1db17598144...
2022-05-23 09:07:11,917 squash.py:37         INFO  docker-squash version 1.0.9, Docker 8728dd2, API 1.41...
2022-05-23 09:07:11,917 squash.py:59         INFO  Using v2 image format
2022-05-23 09:07:11,928 image.py:166        INFO  Old image has 8 layers
2022-05-23 09:07:11,934 image.py:197        INFO  Checking if squashing is necessary...
2022-05-23 09:07:11,935 image.py:207        INFO  Attempting to squash last 6 layers...
2022-05-23 09:07:11,935 image.py:350        INFO  Saving image sha256:f1db175981446fee1736123e371cc491ed66295e9198c5a756cfeae69c34069c to /var/folders/_b/j1r9wbjx0cv8q_ckz9xwp3l80000gn/T/docker-squash-ntwzxqjb/old directory...
2022-05-23 09:07:13,558 image.py:385        INFO  Image saved!
2022-05-23 09:07:13,559 image.py:217        INFO  Squashing image 'f1db17598144'...
2022-05-23 09:07:13,783 image.py:228        INFO  Original image size: 80.01 MB
2022-05-23 09:07:13,783 image.py:229        INFO  Squashed image size: 80.01 MB
2022-05-23 09:07:13,783 image.py:234        INFO  Image size decreased by 0.00 %
2022-05-23 09:07:13,783 squash.py:94         INFO  New squashed image ID is a5c6344a23b3cd0777dc3f748267ac01f3b4f6d6e7eef5cfee5ffee1aa0dcb32
2022-05-23 09:07:15,545 squash.py:81         INFO  Removing old f1db17598144 image...
2022-05-23 09:07:15,574 squash.py:85         INFO  Image f1db17598144 removed!
2022-05-23 09:07:15,574 squash.py:114        INFO  Done
2022-05-23 09:07:15,603 docker_builder.py:265        INFO  Image built and available under following tags: dummy:1, dummy:latest
2022-05-23 09:07:15,604 cli.py:462        INFO  Finished!

 > cekit test behave
2022-05-23 09:07:22,622 cli.py:418        WARNING You are running unreleased development version of CEKit, use it only at your own risk!
2022-05-23 09:07:22,625 tools.py:330        WARNING You are running CEKit on an unknown platform. External dependencies suggestions may not work!
2022-05-23 09:07:22,625 tools.py:500        WARNING The certifi library (https://certifi.io/) was found, depending on the operating system configuration this may result in certificate validation issues
2022-05-23 09:07:22,625 tools.py:506        WARNING Certificate Authority (CA) bundle in use: '/Users/agreen/code/cekit-golden-images/cekit/lib/python3.9/site-packages/certifi/cacert.pem'
2022-05-23 09:07:22,626 base.py:84         INFO  Initializing image descriptor...
2022-05-23 09:07:22,629 resource.py:160        WARNING No value found for 'name' in '{"dest": "/tmp/artifacts/", "path": "./modules"}' artifact; using auto-generated value of 'modules'
2022-05-23 09:07:22,633 resource.py:208        INFO  Copying resource 'modules'...
2022-05-23 09:07:22,640 collector.py:43         INFO  Fetching common steps from 'https://github.com/cekit/behave-test-steps.git'.
Cloning into '/Users/agreen/code/cekit-golden-images/behave-steps-test/target/test'...
remote: Enumerating objects: 14, done.
remote: Counting objects: 100% (14/14), done.
remote: Compressing objects: 100% (13/13), done.
remote: Total 14 (delta 1), reused 7 (delta 0), pack-reused 0
Unpacking objects: 100% (14/14), 12.80 KiB | 374.00 KiB/s, done.
2022-05-23 09:07:23,687 collector.py:77         INFO  Collecting finished!
2022-05-23 09:07:23,687 behave_tester.py:55         WARNING No test collected, test can't be run.
2022-05-23 09:07:23,687 cli.py:462        INFO  Finished!
```

**Environment information:**
 - Operating system: macOS
 - CEKit version: 4.2.0.dev0

**Comment by greenatatlassian** [comment]
Turns out this is deeper than I thought. 

When running in "test" mode, cekit does partially rebuild the target directory. Specifically, `BehaveTester.prepare()` calls `Generator.init()`, which also clears the `target` directory (in addition to the cleanup mentioned in my original report. Then `.init()` recreates some of the `target/` directory, but the modules are only copied in the subsequent function `Generator.generate()`. Adding the latter method call to `BehaveTester.prepare()` fails because we don't have the specific Generator implementation during testing (e.g. `OSBSGenerator` or `DockerGenerator`) and there is a call to an abstract method, `Generator.prepare_artifacts()`.

It is unclear to me from the code what the specific purpose of `init` and `generate` are (i.e. what "belongs" in each), however, moving the call to `copy_modules()` from `generate()` to `init()` seems to solve the issue.

**Comment by rnc** [comment]
Fixed by #784
